source/server/admin/logs_handler.cc

```diff
@@ -29,28 +28,34 @@ absl::flat_hash_map<absl::string_view, spdlog::level::level_enum> buildLevelMap(
 LogsHandler::LogsHandler(Server::Instance& server)
     : HandlerContextBase(server), log_levels_(buildLevelMap()) {}
 
-Http::Code LogsHandler::handlerLogging(absl::string_view url, Http::ResponseHeaderMap&,
-                                       Buffer::Instance& response, AdminStream&) {
-  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+std::vector<absl::string_view> LogsHandler::levelStrings() {
+  std::vector<absl::string_view> strings;
+  strings.reserve(ARRAY_SIZE(spdlog::level::level_string_views));
+  for (spdlog::string_view_t level : spdlog::level::level_string_views) {
+    strings.emplace_back(absl::string_view{level.data(), level.size()});
+  }
+  return strings;
+}
 
-  Http::Code rc = Http::Code::OK;
-  if (!query_params.empty()) {
-    auto status = changeLogLevel(query_params);
-    if (!status.ok()) {
-      rc = Http::Code::BadRequest;
-      response.add(fmt::format("error: {}\n\n", status.message()));
-
-      response.add("usage: /logging?<name>=<level> (change single level)\n");
-      response.add(
-          "usage: /logging?paths=name1:level1,name2:level2,... (change multiple levels)\n");
-      response.add("usage: /logging?level=<level> (change all levels)\n");
-      response.add("levels: ");
-      for (auto level_string_view : spdlog::level::level_string_views) {
-        response.add(fmt::format("{} ", level_string_view));
-      }
+Http::Code LogsHandler::handlerLogging(absl::string_view, Http::ResponseHeaderMap&,
```

Comment:
Is the first parameter used for any handler, since it's now available on the AdminStream? Is this no longer needed on `MAKE_ADMIN_HANDLER`? Also several functions in admin.cc pass `path_and_query` around as arguments. Should all of these just get it from AdminStream?

Reply:
This would require AdminFilter to store the query params once parsed.

Reply:
done as a pending follow-up PR: https://github.com/envoyproxy/envoy/pull/22571 -- it's not reviewable now but once this merges, that one will be non-functional so that its sprawl is easier to reason about.